odoo/addons/base/models/res_users.py

```diff
@@ -199,6 +199,85 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             return len(groups) if count else groups.ids
         return super(Groups, self)._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
 
+    def _get_recursive_implied_groups(self, recursive_groups):
+        """ Returns the implied groups of the group being passed (including the
+            passed group itself)
+        """
+        groups = self.implied_ids.filtered(
+            lambda grp:
+            grp.category_id == self.category_id and
+            grp not in recursive_groups
+        )
+        if groups:
+            recursive_groups |= groups
+            recursive_groups = groups._get_recursive_implied_groups(recursive_groups)
+        return recursive_groups
```

Comment:
I think it's possible to do the same without recursion (and so you don't need the function argument)
```python
def _get_implied_groups(self):
    current_groups = self
    implied_groups = self.env['res.group']
    while current_groups:
        new_implied_groups = current_groups.implied_ids.filtered(
            lambda grp: grp.category_id == self.category_id and grp not in implied_groups
        )
        implied_groups |= new_implied_groups
        current_groups = new_implied_groups
    return implied_groups
```